### PR TITLE
amp-nexxtv-player updates

### PR DESCRIPTION
**Pull request by neko-fire**
# amp-nexxtv-player updates & maintenance

♻️ 
- optimized iframe creation
- added paramter data-exit-mode
- renamed album -> audioalbum
- data-domain can be used now as data-client
- added consent parameter from amp-consent
- added load event

🚮 removed optional parameters: data-origin, data-seek-to, playlist-mask, delay
📖 updated readme and code sample

### CODE REVIEW

**Comment by honeybadgerdontcare** [comment]
@twifkak for validator review

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (alanorozco) [inline extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:195]: It seems like you're only using `consentString_` above. Do you need get the two other consent values?

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (alanorozco) [inline extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:124]: Consider `Element.dataset` and `addParamsToUrl`. This method could clean up nicely:

```js
  getVideoIframeSrc_() {
    const {
      client,
      domainId,
      mediaid,
      mode,
      streamtype,
      disableAds,
      streamingFilter,
      exitMode,
    } = el.dataset;

    const mediaId = userAssert(
      mediaid,
      'The data-mediaid attribute is required for <amp-nexxtv-player> %s'
    );

    const clientId = userAssert(
      client || domainId,
      'One of data-client or data-domain-id attributes is required for <amp-nexxtv-player> %s',
      el
    );

    const url =
      this.origin_ +
      [clientId, streamtype, mediaid].map(encodeURIComponent).join('/');

    return assertAbsoluteHttpOrHttpsUrl(
      addParamsToUrl(
        url,
        dict({
          platform: 'amp',
          disableAds: disableAds,
          streamingFilter: streamingFilter,
          exitMode: exitMode,
          consentString: this.consentString_,
        })
      )
    );
  }
```

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (alanorozco) [inline extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:139]:
```suggestion
      'One of data-client or data-domain-id attributes is required for <amp-nexxtv-player> %s',
```

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (alanorozco) [inline extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:286]: nit: reuse introduced helper

```suggestion
    redispatch(this.element, eventType, {
```

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (alanorozco) [inline extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:52]: @twifkak It's fine to remove `data-` attribute from validation since they're always valid, right?

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (alanorozco) [inline extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:54]: Please bring back `value: "album"` and `value: "playlist-marked"`. Otherwise any existing document that uses those values will be rendered invalid.

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (alanorozco) [inline extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:211]: You should probably add a test confirming that the consent value is passed through.

**Review by alanorozco — CHANGES_REQUESTED** [review]
Thanks @neko-fire

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (twifkak) [reply extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:52]: Yeah, that should be fine. All `data-` attributes are valid by default. This just means it'll no longer be required to contain an http(s) URL; I could contain anything. Since it's ignored in the new code, this shouldn't create any errors.

One question I have is if there are any pages in the wild that use a value other than https://embed.nexx.cloud/ right now. Hopefully there aren't and we can just make the change. But if there are, this seems like it might be a surprising change for them. If the origin is changed to https://embed.next.cloud, will the media still work? Will the site owners' other expectations still be met? (I'm not sure why people would choose a custom origin, so I don't want to guess what those expectations are.)

An alternative is to change the runtime to just show a placeholder if data-origin has been customized. I'm not sure what's better or worse for your users. Just wanted to point that out, make sure you had the chance to make the decision.

Sorry if this was already discussed elsewhere or if I'm overthinking things.

Comment (alanorozco) [reply extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:52]: This is a valid point.

@neko-fire Will documents that set `data-origin` continue to work properly?

Comment (neko-fire) [reply extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:52]: yes we made already sure that this will work as intended with https://embed.next.cloud

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (neko-fire) [reply extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js]: yea you're right, I initially wanted to use, but forgot to remove.

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (neko-fire) [reply extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:107]: I've reworked this part.

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (neko-fire) [reply extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js]: done

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (neko-fire) [reply extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js]: done

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (neko-fire) [reply extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:56]: done

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (neko-fire) [reply extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js:176]: I have added a test, but could not run the test locally since all tests drop Error: Timeout of 2000ms exceeded. I'm not sure how to fix this.

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (twifkak) [inline extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:56]: Please alphabetize lists of values.

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (twifkak) [inline extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii:62]: Ditto.

**Review by twifkak — APPROVED** [review]
Thanks! Looks good for validator changes; left a couple of small requests.

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (neko-fire) [reply extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii]: done

File: extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii
Comment (neko-fire) [reply extensions/amp-nexxtv-player/validator-amp-nexxtv-player.protoascii]: done

**Comment by neko-fire** [comment]
Anything else that needs to be changed?

**Review by alanorozco — COMMENTED** [review]
Sorry @neko-fire, I failed to clicked `submit` on this review previously 😳

File: extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js
Comment (neko-fire) [reply extensions/amp-nexxtv-player/0.1/amp-nexxtv-player.js]: @alanorozco I've changed all your requests regarding consentString. But unit test keeps now failing due to timeout, I don't know how to fix this.

**Comment by neko-fire** [comment]
Any news on this topic @alanorozco

**Review by alanorozco — APPROVED** [review]
Looks good to me, thank you.

**Comment by alanorozco** [comment]
@neko-fire Please ensure that the previous tests pass. I can't merge this otherwise.

These are probably timing out since we're now waiting for a consent response before resolving `layoutCallback`. You should probably mock/stub the consent responses.

```
amp-nexxtv-player
  
  ✗ renders nexxtv video player
Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  ✗ removes iframe after unlayoutCallback
Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  ✗ should forward events from nexxtv-player to the amp element
Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.

  ✗ should pass consent value to iframe
Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
```

https://app.circleci.com/pipelines/github/ampproject/amphtml/4686/workflows/10beb7b2-696f-426d-9ddc-14f4c6ea852c/jobs/61269

**Comment by neko-fire** [comment]
> @neko-fire Please ensure that the previous tests pass. I can't merge this otherwise.
> 
> These are probably timing out since we're now waiting for a consent response before resolving `layoutCallback`. You should probably mock/stub the consent responses.
> 
> ```
> amp-nexxtv-player
>   
>   ✗ renders nexxtv video player
> Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
> 
>   ✗ removes iframe after unlayoutCallback
> Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
> 
>   ✗ should forward events from nexxtv-player to the amp element
> Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
> 
>   ✗ should pass consent value to iframe
> Error: Timeout of 10000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
> ```
> 
> https://app.circleci.com/pipelines/github/ampproject/amphtml/4686/workflows/10beb7b2-696f-426d-9ddc-14f4c6ea852c/jobs/61269

Hi @alanorozco , I've fixed the issue with the test.

**Comment by alanorozco** [comment]
@neko-fire: Since we removed `playerReadyPromise` to fix the test, your player might miss calls to autoplay etc. Did you manualy test with a throttled network?

**Comment by neko-fire** [comment]
> @neko-fire: Since we removed `playerReadyPromise` to fix the test, your player might miss calls to autoplay etc. Did you manualy test with a throttled network?

Yes I do during my frontend test, since the player also checks the connection to play out the right quality. Also our player does not support autoplay in iframe context.

**Comment by neko-fire** [comment]
Hi @alanorozco , do you want me to change something before merge?

**Comment by alanorozco** [comment]
Ok, if it doesn't do autoplay then we're probably fine. Thanks for the update.

**Pull request merged**
